### PR TITLE
Add unit tests for ASMBert with tied embeddings and safe serialization

### PR DESCRIPTION
**Pull request by ranieri**
Note: `save_pretrained()` default changed somewhere between HF Transformers version 4.34 and 4.38 to prefer safe serialisation. This serialization method cannot deal with our way of sharing weights, so we can only currenltly save using explicit `safe_serialization=False`. 

This PR codifies existing behaviour only.

### CODE REVIEW

File: asmtransformers/tests/test_asmbert.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+import torch
+import pytest
+from transformers import BertConfig
+
+from asmtransformers.models.asmbert import ASMBertForMaskedLM
+
+
+def create_model():
```
Comment (akaIDIOT) [inline asmtransformers/tests/test_asmbert.py:8]: This should be a fixture, I'd think?

File: asmtransformers/tests/test_asmbert.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    torch.manual_seed(0)
```
Comment (akaIDIOT) [inline asmtransformers/tests/test_asmbert.py:12]: If this is really needed, I'd expect a note on what's getting randomized that needs to be stabilized here.

File: asmtransformers/tests/test_asmbert.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    assert position_embeddings.weight.data_ptr() == word_embeddings.weight.data_ptr()
+
+
+def test_forward_reports_losses_and_jtp_ignores_out_of_range_labels():
+    model = create_model()
+    model.eval()
+
+    input_ids = torch.tensor([[1, 2, 3, 4]])
+    labels = torch.tensor([[7, 9, -100, 31]])
+    labels_with_different_non_jtp_targets = torch.tensor([[7, 10, -100, 30]])
```
Comment (akaIDIOT) [inline asmtransformers/tests/test_asmbert.py:45]: The intention here is not very clear. The actual code was a while ago, but what does labels imply here? 🤔 

The second set of labels is also not very different; does the name mean that it's just different, and still no jump targets? Or does the first contain jump targets but the second one does not?

File: asmtransformers/tests/test_asmbert.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    optimizer.step()
+
+    position_weight = model.base_model.embeddings.position_embeddings.weight
+
+    assert position_weight is shared_weight
+    assert position_weight.data_ptr() == shared_weight.data_ptr()
+    assert not torch.equal(shared_weight_before, shared_weight.detach())
+
+
+def test_pickle_save_and_load_preserves_tied_embeddings(tmp_path):
```
Comment (akaIDIOT) [inline asmtransformers/tests/test_asmbert.py:91]: I like the simplicity of this test, create-save-load-check 👏🏻 (and an important check at that).

File: asmtransformers/tests/test_asmbert.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+def test_safe_serialization_rejects_tied_embeddings(tmp_path):
+    model = create_model()
+
+    # The embeddings are tied at runtime, but that tie is not declared through
+    # Transformers' tied-weight metadata. Safe serialization should therefore fail.
+    # See:
+    # https://huggingface.co/docs/transformers/main_classes/model#transformers.PreTrainedModel.get_expanded_tied_weights_keys
+    #
+    # If the model starts declaring this tie properly, replace this with a positive
+    # safe-serialization round-trip test.
```
Comment (akaIDIOT) [inline asmtransformers/tests/test_asmbert.py:119]: Fair enough. Would it make sense to override `save_pretrained` without the parameter and call super with `safe_serialization=False`? Might be overkill for a test that is intended as a regression for the upgrade in #20 (?).

**Review by akaIDIOT — APPROVED** [review]

**Comment by ranieri** [comment]
Lol doe es niet werken zo laat, man!

File: asmtransformers/tests/test_asmbert.py (hunk 1 above)
Comment (ranieri) [reply asmtransformers/tests/test_asmbert.py:8]: I considered it. For this particular set of tests, I like to start fresh every time.

File: asmtransformers/tests/test_asmbert.py (hunk 3 above)
Comment (drxzcl) [reply asmtransformers/tests/test_asmbert.py:45]: The pair contains the same jump targets (<8, in the model instantiated up top), but different non-jump targets. The test checks that the JTP loss is the same, but the total loss is different.

This might need a line comment :P

File: asmtransformers/tests/test_asmbert.py (hunk 5 above)
Comment (drxzcl) [reply asmtransformers/tests/test_asmbert.py:119]: In an ordinary world, yes. But unsafe save is getting the same guard in ST5, so we need to do it properly anyway.

File: asmtransformers/tests/test_asmbert.py (hunk 2 above)
Comment (drxzcl) [reply asmtransformers/tests/test_asmbert.py:12]: Added comment

File: asmtransformers/tests/test_asmbert.py (hunk 1 above)
Comment (akaIDIOT) [reply asmtransformers/tests/test_asmbert.py:8]: That's ... how a fixture works 🤔 

Unless you tell it to stick around per module / session / ...
